game-core/src/main/java/games/strategy/engine/history/change/units/TransformDamagedUnits.java

```diff
@@ -0,0 +1,127 @@
+package games.strategy.engine.history.change.units;
+
+import games.strategy.engine.data.Change;
+import games.strategy.engine.data.CompositeChange;
+import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitType;
+import games.strategy.engine.data.changefactory.ChangeFactory;
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.engine.history.change.HistoryChange;
+import games.strategy.triplea.UnitUtils;
+import games.strategy.triplea.attachments.UnitAttachment;
+import games.strategy.triplea.formatter.MyFormatter;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import lombok.AccessLevel;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.Value;
+import lombok.experimental.FieldDefaults;
+import org.triplea.util.Tuple;
+
+/**
+ * Transforms units into other unit types as determined by {@link
+ * UnitAttachment#getWhenHitPointsDamagedChangesInto()}
+ */
+@FieldDefaults(makeFinal = true, level = AccessLevel.PRIVATE)
+@Getter
+@EqualsAndHashCode
+public class TransformDamagedUnits implements HistoryChange {
+
+  CompositeChange change = new CompositeChange();
+  Territory location;
+  /** Map of old unit -> new unit */
+  Map<Unit, Unit> transformingUnits = new HashMap<>();
+
+  CompositeChange attributeChanges = new CompositeChange();
+
+  public TransformDamagedUnits(final Territory location, final Collection<Unit> damagedUnits) {
+    this.location = location;
+
+    // check if each of the damaged units are supposed to change when they take damage
+    // if it is supposed to change, create the new unit and translate attributes from the old unit
+    // to the new unit
+    for (final Unit unit : damagedUnits) {
+      final Map<Integer, Tuple<Boolean, UnitType>> map =
```

Comment:
side-note, we're trying to kill `Tuple` and replace it with value objects. For example `TotalPowerAndRolls` used to be a `Tuple<Integer,Integer>`,that conversion certainly lead to more legible code. In this case, the `Boolean` is very cryptic, it's kinda hard to know what the represents at first glance.